File: mcp-code-intelligence-python/tests/test_port_conflict.py
"""Port conflict bug fix tests — ViewerServer port release on shutdown.

Bug: When MCP server is reconnected, old instance doesn't release the HTTP
viewer port, causing "port already in use" error.

Fix: ViewerServer.stop() calls HTTPServer.shutdown() which releases the port.
McpServer._shutdown() calls self._viewer.stop() on stdin close + SIGTERM/SIGINT.
"""

import socket
import sys
import time
import threading
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from mcp_code_intel.http.viewer_server import ViewerServer


def _find_free_port() -> int:
    """Find an available port for testing."""
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("127.0.0.1", 0))
        return s.getsockname()[1]


def _is_port_in_use(port: int) -> bool:
    """Check if a port is currently bound."""
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        return s.connect_ex(("127.0.0.1", port)) == 0


def test_viewer_stop_releases_port():
    """Unit test: ViewerServer.stop() closes HTTP server and releases port."""
    port = _find_free_port()
    server = ViewerServer(port, ".")
    server.start()
    time.sleep(0.3)

    # Port should be in use
    assert _is_port_in_use(port), f"Port {port} should be bound after start()"

    # Stop should release port
    server.stop()
    time.sleep(0.3)

    assert not _is_port_in_use(port), f"Port {port} should be free after stop()"


def test_start_stop_start_same_port():
    """Reconnect simulation: start → stop → start new instance on same port."""
    port = _find_free_port()

    # First instance
    server1 = ViewerServer(port, ".")
    server1.start()
    time.sleep(0.3)
    assert _is_port_in_use(port)

    # Simulate stdin close → shutdown
    server1.stop()
    time.sleep(0.3)
    assert not _is_port_in_use(port)

    # Second instance on same port — should succeed (no "port in use" error)
    server2 = ViewerServer(port, ".")
    server2.start()
    time.sleep(0.3)
    assert _is_port_in_use(port), "Second instance should bind same port"

    # Cleanup
    server2.stop()
    time.sleep(0.3)
    assert not _is_port_in_use(port)


def test_stop_idempotent():
    """Calling stop() multiple times should not raise errors."""
    port = _find_free_port()
    server = ViewerServer(port, ".")
    server.start()
    time.sleep(0.3)

    server.stop()
    server.stop()  # Should not raise
    server.stop()  # Should not raise


def test_stop_without_start():
    """Calling stop() before start() should not raise errors."""
    port = _find_free_port()
    server = ViewerServer(port, ".")
    server.stop()  # Should not raise — _server is None


if __name__ == "__main__":
    print("Running port conflict tests...")
    test_viewer_stop_releases_port()
    print("  PASS: test_viewer_stop_releases_port")
    test_start_stop_start_same_port()
    print("  PASS: test_start_stop_start_same_port")
    test_stop_idempotent()
    print("  PASS: test_stop_idempotent")
    test_stop_without_start()
    print("  PASS: test_stop_without_start")
    print("\n4 passed, 0 failed")

File: mcp-code-intelligence-nodejs/src/test-port-conflict.ts
/**
 * Port conflict bug fix tests — ViewerServer port release on shutdown.
 *
 * Bug: When MCP server is reconnected, old instance doesn't release the HTTP
 * viewer port, causing "port already in use" error.
 *
 * Fix: ViewerServer.stop() calls server.close() which releases the port.
 * cleanup() in index.ts calls _viewerServer.stop() on stdin close + SIGTERM/SIGINT.
 *
 * Run: node --experimental-specifier-resolution=node dist/test-port-conflict.js
 * (after: npx tsc)
 */

import * as net from 'net';
import * as http from 'http';
import { ViewerServer } from './http/viewer-server.js';

let passed = 0;
let failed = 0;

function findFreePort(): Promise<number> {
  return new Promise((resolve) => {
    const srv = net.createServer();
    srv.listen(0, '127.0.0.1', () => {
      const port = (srv.address() as net.AddressInfo).port;
      srv.close(() => resolve(port));
    });
  });
}

function isPortInUse(port: number): Promise<boolean> {
  return new Promise((resolve) => {
    const client = new net.Socket();
    client.once('connect', () => { client.destroy(); resolve(true); });
    client.once('error', () => { client.destroy(); resolve(false); });
    client.connect(port, '127.0.0.1');
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function testStopReleasesPort(): Promise<void> {
  const port = await findFreePort();
  const server = new ViewerServer(port, '.');
  server.start();
  await sleep(300);

  const bound = await isPortInUse(port);
  if (!bound) { failed++; console.error('  FAIL: port not bound after start'); return; }

  server.stop();
  await sleep(300);

  const free = !(await isPortInUse(port));
  if (!free) { failed++; console.error('  FAIL: port not released after stop'); return; }

  passed++;
  console.log('  PASS: test_stop_releases_port');
}

async function testReconnectSamePort(): Promise<void> {
  const port = await findFreePort();

  // First instance
  const server1 = new ViewerServer(port, '.');
  server1.start();
  await sleep(300);

  if (!(await isPortInUse(port))) { failed++; console.error('  FAIL: server1 not bound'); return; }

  // Simulate stdin close → cleanup
  server1.stop();
  await sleep(300);

  if (await isPortInUse(port)) { failed++; console.error('  FAIL: port not released after stop'); return; }

  // Second instance on same port — should succeed
  const server2 = new ViewerServer(port, '.');
  try {
    server2.start();
    await sleep(300);
    if (!(await isPortInUse(port))) { failed++; console.error('  FAIL: server2 not bound'); return; }
    server2.stop();
    await sleep(300);
  } catch (e) {
    failed++;
    console.error(`  FAIL: second instance threw: ${e}`);
    return;
  }

  passed++;
  console.log('  PASS: test_reconnect_same_port');
}

async function testStopIdempotent(): Promise<void> {
  const port = await findFreePort();
  const server = new ViewerServer(port, '.');
  server.start();
  await sleep(300);

  try {
    server.stop();
    server.stop(); // Should not throw
    server.stop(); // Should not throw
  } catch (e) {
    failed++;
    console.error(`  FAIL: multiple stop() threw: ${e}`);
    return;
  }

  passed++;
  console.log('  PASS: test_stop_idempotent');
}

async function testStopWithoutStart(): Promise<void> {
  const port = await findFreePort();
  const server = new ViewerServer(port, '.');

  try {
    server.stop(); // Should not throw — server is null
  } catch (e) {
    failed++;
    console.error(`  FAIL: stop() without start() threw: ${e}`);
    return;
  }

  passed++;
  console.log('  PASS: test_stop_without_start');
}

async function main(): Promise<void> {
  console.log('Running Node.js port conflict tests...');
  await testStopReleasesPort();
  await testReconnectSamePort();
  await testStopIdempotent();
  await testStopWithoutStart();
  console.log(`\n${passed} passed, ${failed} failed`);
  process.exit(failed > 0 ? 1 : 0);
}

main();
